magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamController.scala

```diff
@@ -0,0 +1,114 @@
+                      (implicit val system: ActorSystem,
+                       implicit val config: Config,
+                       implicit val materializer: Materializer) {
+
+  implicit val ec: ExecutionContextExecutor = system.dispatcher
+  implicit val scheduler: Scheduler = system.scheduler
+
+  val log = Logging(system, getClass)
+
+  private val ssc = new StreamSourceController(bufferSize, this)
+  private val (actorRef, source) = ssc.refAndSource
+  private val crawledCount = new AtomicLong(0)
+  private var tokenOptionF: Future[Option[String]] = Future{None}
+
+  private def getDataSets(nextFuture: () => Future[(Option[String], List[DataSet])])
+  : Future[(Option[String], List[DataSet])] = {
+
+    val onRetry = (retryCount: Int, e: Throwable) =>
+      log.error(e, "Failed while fetching from registry, retries left: {}", retryCount + 1)
+
+    val safeFuture: Future[(Option[String], List[DataSet])] =
+      ErrorHandling.retry(nextFuture, 30.seconds, 30, onRetry)
+        .recover {
+          case e: Throwable =>
+            log.error(e, "Failed completely while fetching from registry. " +
+              "This means we can't go any further!!")
+            (None, Nil)
+        }
+
+    safeFuture
+  }
+
+  private def fillStreamSource(nextFuture: () => Future[(Option[String], List[DataSet])],
+                               isFirst: Boolean = false)
+  : Future[Option[String]] = {
+
+    getDataSets(nextFuture)
+      .map(results => {
+        val tokenOption = results._1
+        val dataSets = results._2
+        crawledCount.addAndGet(dataSets.size)
+        log.info("Total crawled {} datasets from registry", crawledCount.get())
+        val hasNext = tokenOption.nonEmpty && dataSets.nonEmpty
+        ssc.fillSource(dataSets, hasNext, isFirst)
+        tokenOption
+      })
+  }
+
+  /**
+    * The returned source is used to construct a dataset stream.
+    *
+    * @return the stream source
+    */
+  def getSource: Source[DataSet, NotUsed] = {
+    source
+  }
+
+  /**
+    * Get the first batch datasets from registry and fill the stream source.
+    *
+    * @return future optional token. If nonempty, indicate more datasets are available.
+    */
+  def start(): Future[Option[String]] = {
+    val firstPageF = () => interface.getDataSetsReturnToken(start = 0, size = bufferSize)
+    tokenOptionF = fillStreamSource(firstPageF, isFirst = true)
+    tokenOptionF
+  }
+
+  /**
+    * Get the next batch datasets from registry and fill the stream source.
+    *
+    * @param nextSize the max size of the batch
+    * @return future optional token. If nonempty, indicate more datasets are available.
+    */
+  def next(nextSize: Int): Future[Option[String]] = {
+    tokenOptionF.flatMap(tokenOption => {
+      if (tokenOption.isEmpty){
+        Future.successful(None)
+      }
+      else {
+        val nextPageF = () => interface.getDataSetsToken(tokenOption.get, bufferSize/2)
```

Comment:
Tested locally via minikube cluster.
Seems indexer will stuck on this line above and terminal printed the following error:
```
[ERROR] [12/03/2018 13:42:42.360] [indexer-akka.actor.default-dispatcher-34] [akka://indexer/user/StreamSupervisor-0/flow-203-0-unknown-operation] Error during preStart in [akka.http.scaladsl.model.HttpEntity$Limitable@7eb90a75]
indexer EntityStreamSizeException: actual entity size (Some(10015105)) exceeded content length limit (8388608 bytes)! You can configure this by setting `akka.http.[server|client].parsing.max-content-length` or calling `HttpEntity.withSizeLimit` before materializing the dataBytes stream.
indexer         at akka.http.scaladsl.model.HttpEntity$Limitable$$anon$1.preStart(HttpEntity.scala:607)
indexer         at akka.stream.impl.fusing.GraphInterpreter.init(GraphInterpreter.scala:520)
indexer         at akka.stream.impl.fusing.GraphInterpreterShell.init(ActorGraphInterpreter.scala:380)
indexer         at akka.stream.impl.fusing.ActorGraphInterpreter.tryInit(ActorGraphInterpreter.scala:538)
indexer         at akka.stream.impl.fusing.ActorGraphInterpreter.preStart(ActorGraphInterpreter.scala:586)
indexer         at akka.actor.Actor$class.aroundPreStart(Actor.scala:510)
indexer         at akka.stream.impl.fusing.ActorGraphInterpreter.aroundPreStart(ActorGraphInterpreter.scala:529)
indexer         at akka.actor.ActorCell.create(ActorCell.scala:590)
indexer         at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:461)
indexer         at akka.actor.ActorCell.systemInvoke(ActorCell.scala:483)
indexer         at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)
indexer         at akka.dispatch.Mailbox.run(Mailbox.scala:223)
indexer         at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
indexer         at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
indexer         at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
indexer         at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
indexer         at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```
It looks it tries to request registry for 500 records in one single HTTP request --- It's probably too big...
I think no larger than 50 records in a single HTTP request would be reasonable (before was hard coded to 50 I think)

Reply:
The size is configurable in appliation.conf. We can reduce it.

Reply:
The problem is that if we change `streamControllerSourceBufferSize` = 100 then crawler will never prepare more than 100 records while elasticsearch processes bulk index request. In the meantime, one elasticsearch bulk request normally consumes 500 ~ 1000 records.

Reply:
I successfully ran a local test with setting streamControllerSourceBufferSize = 100. 
This parameter is independent of elasticsearch bulk indexing size.

Reply:
For this one, we will be fine as long we change `streamControllerSourceBufferSize` to 100:
```
crawler {
  streamControllerSourceBufferSize = 100
}
```